### DISCUSSION TITLE
Using Android CLI Tools

### DISCUSSION THREAD

**Discussion by DaveBolt** [ionic-team/capacitor#3224] [Q&A] [answered]
Today I had to build a project that uses Capacitor instead of Cordova for the first time.
I finally figured out a command to build the project, not necessarily the best way to do it: ionic capacitor build android
This failed because it couldn't find my Android Studio installation.
After searching my entire drive I finally found the path to studio.sh ("/snap/android-studio/current/android-studio/bin/studio.sh" on my Ubuntu 18.04 installation), and added it to capacitor.config.json.
Re-running the build, I ended up in Android Studio (took a long time to open), where I then had to manually trigger a build of the final app.
If this had been a Cordova build, the entire process including the final build of the app would have run automatically.
Now, I finally find the documentation about installing Capacitor, with the mysterious comment on this page,
https://capacitorjs.com/docs/basics/building-your-app
"Android relies on Android Studio (or, optionally, the Android CLI tools) to build the app:"
Unfortunately, no further information seems to be available.
A Cordova build uses Android Studio CLI tools, as far as I can tell, and also uses the defined environment variables to the Android SDK.
Can I do something similar with a Capacitor project? If so, where is the documentation.

**Comment by imhoffd** [comment, marked as answer]
For [Capacitor 3](https://github.com/ionic-team/capacitor/issues/3133), we plan to offer a `npx cap run` command that provides a CLI workflow for deploying **debug** builds to devices, intended to be used during development.

We will **not** be adding support for **release** builds on the command-line at this time. You can follow [this issue](https://github.com/ionic-team/capacitor/issues/324) for that, if that's what you're asking for. I don't think Capacitor is the right tool to handle release builds because:

- Cordova offers programmatic builds because it attempts to completely abstract away the native projects, whereas Capacitor embraces them. With Capacitor, you are encouraged to commit native projects to source control and tweak them as you see fit. Read about more [differences between Capacitor and Cordova](https://capacitorjs.com/docs/cordova#differences-between-capacitor-and-cordova) in the docs. 
- Native project builds, especially release builds, are complex and constantly in flux. It would take great effort to take the Cordova approach here to abstract away this process. The native projects in Capacitor apps are just native apps. Open them with their respective native IDEs and you can do anything you want to them. Android Studio and Xcode offer a graphical interface for navigating these complex builds, and we encourage you to use them.
- Since the native projects in Capacitor apps are just native apps, any existing tooling for native apps can be used. For example, [Fastlane](https://fastlane.tools/) can be used locally or in CI servers to manage signing and even app store deployment.

**Reply by DaveBolt** [reply]
Nice clear reply.
So to be clear
If the project is using Capacitor, I progress to the point of building the app for either Android or iOS, then I use Android Studio or xCode, as appropriate to do the final build, incorporating tweaks as relevant for the specific environments.
I can live with that.
Would be good though, if for instance the path to studio.sh could be in an environment variable. With the current setup I put the path into the capacitor.config.json file, commit to Git, then another developer gets a problem because he is using Windows not Ubuntu and Android Studio is on a completely different path. Or am I (not surprisingly), missing something else?

**Reply by imhoffd** [reply]
Yes, correct! 

As for the path to Android Studio, we are already planning to separate "system settings" from "app settings" for our Capacitor 3 configuration project. You are correct, it's not great to have user/system settings in a committed file. You can subscribe to https://github.com/ionic-team/capacitor/issues/3133 for progress.

Thanks for validating our thoughts! 💙
